controller_manager/test/test_hardware_spawner.cpp

```diff
@@ -0,0 +1,282 @@
+  : ControllerManagerFixture<controller_manager::ControllerManager>(""),
+    RMServiceCaller(TEST_CM_NAME)
+  {
+    cm_->set_parameter(rclcpp::Parameter(
+      "hardware_components_initial_state.unconfigured",
+      std::vector<std::string>{"TestSystemHardware"}));
+  }
+
+public:
+  void SetUp() override
+  {
+    ControllerManagerFixture::SetUp();
+
+    update_timer_ = cm_->create_wall_timer(
+      std::chrono::milliseconds(10),
+      [&]()
+      {
+        cm_->read(time_, PERIOD);
+        cm_->update(time_, PERIOD);
+        cm_->write(time_, PERIOD);
+      });
+
+    update_executor_ =
+      std::make_shared<rclcpp::executors::MultiThreadedExecutor>(rclcpp::ExecutorOptions(), 2);
+
+    update_executor_->add_node(cm_);
+    update_executor_spin_future_ =
+      std::async(std::launch::async, [this]() -> void { update_executor_->spin(); });
+    // This sleep is needed to prevent a too fast test from ending before the
+    // executor has began to spin, which causes it to hang
+    std::this_thread::sleep_for(50ms);
+  }
+
+  void TearDown() override { update_executor_->cancel(); }
+
+  rclcpp::TimerBase::SharedPtr robot_description_sending_timer_;
+
+protected:
+  rclcpp::TimerBase::SharedPtr update_timer_;
+
+  // Using a MultiThreadedExecutor so we can call update on a separate thread from service callbacks
+  std::shared_ptr<rclcpp::Executor> update_executor_;
+  std::future<void> update_executor_spin_future_;
+};
+
+TEST_F(TestHardwareSpawnerWithoutRobotDescription, when_no_robot_description_spawner_times_out)
+{
+  EXPECT_EQ(
+    call_spawner(
+      "TestSystemHardware --configure -c test_controller_manager --controller-manager-timeout 1.0"),
+    256)
+    << "could not change hardware state because not robot description and not services for "
+       "controller "
+       "manager are active";
+}
+
+TEST_F(TestHardwareSpawnerWithoutRobotDescription, spawner_with_later_load_of_robot_description)
+{
+  // Delay sending robot description
+  robot_description_sending_timer_ = cm_->create_wall_timer(
+    std::chrono::milliseconds(2500), [&]() { pass_robot_description_to_cm_and_rm(); });
+
+  EXPECT_EQ(
+    call_spawner(
+      "TestSystemHardware --configure -c test_controller_manager --controller-manager-timeout 1.0"),
+    256)
+    << "could not activate control because not robot description";
+  EXPECT_EQ(
+    call_spawner(
+      "TestSystemHardware --configure -c test_controller_manager --controller-manager-timeout 2.5"),
+    0);
+  EXPECT_EQ(
+    get_component_state("TestSystemHardware").id,
+    lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE);
+}
+
+class TestHardwareSpawnerWithNamespacedCM
+: public ControllerManagerFixture<controller_manager::ControllerManager>,
+  public RMServiceCaller
+{
+public:
+  TestHardwareSpawnerWithNamespacedCM()
+  : ControllerManagerFixture<controller_manager::ControllerManager>(
+      ros2_control_test_assets::minimal_robot_urdf, "foo_namespace"),
```

Comment:
```suggestion
      ros2_control_test_assets::minimal_robot_urdf, false, "foo_namespace"),
```

This might be needed due to API here : https://github.com/ros-controls/ros2_control/blob/a7988ad9190e14f6fc47fbac97892f6619dd1477/controller_manager/test/controller_manager_test_common.hpp#L67-L69